### PR TITLE
[BEAM-720] Enable WindowedWordCountIT on Flink runner in presubmit

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by kennknowles** [comment]
@aljoscha when I ran this locally with the in-process Flink it passed, so we might as well have it enabled to try to get some more coverage. Would be nice to have actual Flink cluster coverage of this.

**Comment by kennknowles** [comment]
Also: if I understand correctly, `--streaming=true` still matters for FlinkRunner, so we should have two executions here, right?

**Comment by aljoscha** [comment]
tl;dr: LGTM! 

Yes, the `--streaming` flag is considered but if you omit it the Flink runner will choose batch/streaming based on the whether there are any unbounded sources or not. `WindowedWordCountIT` would be executed as batch since there are no unbounded source.

However, even when you execute this job using `--streaming=true` it will still terminate because we only have bounded sources and the Flink streaming runner will terminate once all operators finish.

I would like to have a proper "streaming" test with unbounded sources where we manually cancel the job once the stopping condition is met. (Like the `TestDataflowRunner` does by waiting for the `PAssert` metrics to show that all PAsserts passed.) For this, however, we need metrics support in the Flink Runner and we need to extend the `TestFlinkRunner` in a similar way. By the way, how does the Dataflow runner know when the `WindowedWordCountIT` has finished?

Also, the success matcher set in `WindowedWordCountIT` is ignored by the `TestFlinkRunner`.

In the end, this still LGTM! Because some coverage is better than none.

**Comment by kennknowles** [comment]
Since you mention a proper "streaming" test, this is most easy to do only with triggers that yield deterministic output, unless a runner implements `TestStream` which offers control over watermarks and processing time, and waiting for quiescence. The Spark runner has `CreateStream` which is almost the same but tweaked for Spark and we should unify those. For Dataflow, my thought was to try to implement `TestStream` fairly soon. For Flink, what do you think?

**Comment by aljoscha** [comment]
I think it will be quite hard to do for Flink because there is no upstream communication between operators. Maybe we can do it via instrumentation (via aggregators) and listening to them. How will it work in the Dataflow Runner?

**Comment by kennknowles** [comment]
I don't have a plan yet, but I'll let you know what I learn :-)

**Comment by kennknowles** [comment]
R: @jasonkuster @aljoscha 

Officially tagging with JIRA and reviewers. Sorry for the mix-up.

**Comment by kennknowles** [comment]
retest this please

**Comment by jasonkuster** [comment]
Jenkins passed and inspection of the logs confirms windowedwordcount executed.

LGTM

**Comment by davorbonaci** [comment]
Is this merge-ready?

**Comment by jasonkuster** [comment]
My understanding was yes.

**Comment by kennknowles** [comment]
Yes. I will merge shortly.
